File: Emulator/src/main/core/net/event/MSWriteEvent.java
package core.net.event;

import core.net.IoSession;
import core.net.IoWriteEvent;
import core.net.packet.IoBuffer;

import java.nio.ByteBuffer;

/**
 * The type Ms write event.
 */
public final class MSWriteEvent extends IoWriteEvent {

    /**
     * Instantiates a new Ms write event.
     *
     * @param session the session
     * @param context the context
     */
    public MSWriteEvent(IoSession session, Object context) {
        super(session, context);
    }

    @Override
    public void write(IoSession session, Object context) {
        IoBuffer b = (IoBuffer) context;
        int size = b.toByteBuffer().position();
        ByteBuffer buffer = ByteBuffer.allocate(1 + size + b.getHeader().ordinal());
        buffer.put((byte) b.opcode());
        switch (b.getHeader()) {
            case NORMAL:
                break;
            case BYTE:
                buffer.put((byte) size);
                break;
            case SHORT:
                buffer.putShort((short) size);
                break;
        }
        buffer.put((ByteBuffer) b.toByteBuffer().flip());
        session.queue((ByteBuffer) buffer.flip());
    }

}

File: game-server/src/main/core/net/event/MSWriteEvent.kt
package core.net.event

import core.net.IoSession
import core.net.IoWriteEvent
import core.net.packet.IoBuffer
import core.net.packet.PacketHeader
import java.nio.ByteBuffer

/**
 * Handles management server write events.
 *
 * @author Emperor
 */
class MSWriteEvent(
    session: IoSession,
    context: Any,
) : IoWriteEvent(session, context) {

    override fun write(session: IoSession, context: Any) {
        val b = context as IoBuffer
        val size = b.toByteBuffer().position()
        val header = b.header ?: PacketHeader.NORMAL
        val buffer = ByteBuffer.allocate(1 + size + header.ordinal)
        buffer.put(b.opcode().toByte())
        when (header) {
            PacketHeader.NORMAL -> {
            }
            PacketHeader.BYTE -> {
                buffer.put(size.toByte())
            }
            PacketHeader.SHORT -> {
                buffer.putShort(size.toShort())
            }
            else -> {
            }
        }
        buffer.put(b.toByteBuffer().flip() as ByteBuffer)
        buffer.flip()
        session.queue(buffer)
    }
}
